### PR TITLE
python3Packages.gradio: 3.20.1 -> 3.43.1

### PR DESCRIPTION
**Pull request by pbsds**
## Description of changes

Cyclic dependencies are fun, and the gradio devs sure can't tag their releases nor provide all needed test artifacts. But thankfully they've reduced the size of their testing closure tremendously. 

Superseeds #254011

~~This does not build in its current state, but sure am i getting close!~~
I was hit by [this issue](https://discourse.nixos.org/t/overriding-docheck-doesnt-work-with-python-package/14674). TIL about `overridePythonAttrs`.

Tests for `gradio-clients`:
image

Tests for `gradio`:
image

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by pbsds** [comment]
@ofborg build python310Packages.gradio python311Packages.gradio python310Packages.gradio-client python311Packages.gradio-client

**Comment by happysalada** [comment]
we have got 12 test failures on darwin https://logs.ofborg.org/?key=nixos/nixpkgs.254041&attempt_id=59d85d89-6d74-4eeb-87b6-21434480c6b6
should we just conditionally disable those ?

File: pkgs/development/python-modules/gradio/default.nix [hunk 1]
```diff
@@ -156,12 +151,16 @@ buildPythonPackage rec {
     # makes pytest freeze 50% of the time
     "test/test_interfaces.py"
   ];
-  #pytestFlagsArray = [ "-x" "-W" "ignore" ]; # uncomment for debugging help
+  pytestFlagsArray = [
+    "-x"  # abort on first failiure
```
Comment (happysalada) [inline pkgs/development/python-modules/gradio/default.nix:155]:
```suggestion
    "-x"  # abort on first failure
```

File: pkgs/development/python-modules/gradio/default.nix [hunk 2]
```diff
@@ -79,55 +78,52 @@ buildPythonPackage rec {
   ];
 
   propagatedBuildInputs = [
-    setuptools # needs pkg_resources
+    setuptools # needed for 'pkg_resources'
```
Comment (happysalada) [inline pkgs/development/python-modules/gradio/default.nix:81]: should we keep dependency order the same as in the requirement.txt , just to make it easier to spot updates ?
https://github.com/gradio-app/gradio/blob/main/requirements.txt

**Comment by happysalada** [comment]
I've left a couple of comments, this is great work!

File: pkgs/development/python-modules/gradio/default.nix (hunk 2 above)
Comment (pbsds) [reply pkgs/development/python-modules/gradio/default.nix:81]: Aside from setuptools it should be that same?

**Comment by pbsds** [comment]
Seems like the build failure is watchfiles, not gradio

File: pkgs/development/python-modules/gradio/default.nix (hunk 2 above)
Comment (happysalada) [reply pkgs/development/python-modules/gradio/default.nix:81]: right , I had only looked before the change. Your change fixed this.

**Comment by happysalada** [comment]
This is ready to merge then. Do you want to fix the type mabye ?

**Comment by happysalada** [comment]
(I just tested on master on a darwin machine and watchfile tests pass. So I'm going to let go of the watchfile error).

File: pkgs/development/python-modules/gradio/default.nix (hunk 2 above)
Comment (pbsds) [reply pkgs/development/python-modules/gradio/default.nix:81]: They sorted it upstream, so I followed suit

**Comment by happysalada** [comment]
I'm planning to merge this tomorrow morning , just to give a bit of time in case anyone else wants to have a look at this.

**Pull request merged**